File: backend/app/core/upgrades/stat_modifiers/explosion_radius.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class ExplosionRadius(UpgradeImplementation):
    id = 'explosion_radius'
    name = 'Blast Radius'
    description = '+5 explosion radius'
    rarity = 'uncommon'
    type = 'stat_modifier'
    cost = 6

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 5
        appliers.modifiers.add_multiplier('bullet', 'explosionRadius', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/explosion_radius.ts
import { Upgrade, type UpgradeDef, type ExplosionSpec } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const ExplosionRadiusDef: UpgradeDef = {
  id: "explosion_radius",
  name: "Blast Radius",
  description: "+5 explosion radius",
  rarity: RarityID.Uncommon,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 6,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.ExplosionRadius,
  value: 5,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 5,
  dependentOn: ["explosive_bullets", "explosion_on_kill"],
  dependencyCount: 1,
}

export class ExplosionRadius extends Upgrade {
  // Stat fields in the def are UI/backend metadata only — the behavior is the
  // hook below, so the default modifier-channel application is skipped.
  onApply(): void {}

  modifyExplosion(explosion: ExplosionSpec): void {
    explosion.radius += this.def.value!
  }
}
